Adding Involved People and Musician Credits frames

NCrusher74 opened the pull request:
**ID3.org spec for version 2.2:**
```
4.4.   Involved people list

   Since there might be a lot of people contributing to an audio file in
   various ways, such as musicians and technicians, the 'Text
   information frames' are often insufficient to list everyone involved
   in a project. The 'Involved people list' is a frame containing the
   names of those involved, and how they were involved. The body simply
   contains a terminated string with the involvement directly followed by
   a terminated string with the involvee followed by a new involvement
   and so on. There may only be one "IPL" frame in each tag.

     Involved people list   "IPL"
     Frame size             $xx xx xx
     Text encoding          $xx
     People list strings    <textstrings>
```

**ID3.org spec for version 2.3:**
```
4.4.   Involved people list

   Since there might be a lot of people contributing to an audio file in
   various ways, such as musicians and technicians, the 'Text
   information frames' are often insufficient to list everyone involved
   in a project. The 'Involved people list' is a frame containing the
   names of those involved, and how they were involved. The body simply
   contains a terminated string with the involvement directly followed
   by a terminated string with the involvee followed by a new
   involvement and so on. There may only be one "IPLS" frame in each
   tag.

     <Header for 'Involved people list', ID: "IPLS">
     Text encoding          $xx
     People list strings    <text strings according to encoding>
```

**ID3.org spec for version 2.4:**
```
  TMCL
   The 'Musician credits list' is intended as a mapping between
   instruments and the musician that played it. Every odd field is an
   instrument and every even is an artist or a comma delimited list of
   artists.

  TIPL
   The 'Involved people list' is very similar to the musician credits
   list, but maps between functions, like producer, and names.
```

NCrusher74 committed:
Adding Involved People and Musician Credits frames

SDGGiesbrecht commented:
Do you have a specific question? Those specifications already tell you exactly what to do.

NCrusher74 commented:
Sorry, I was getting to that and then got distracted. Mostly I'm just setting up because I know I'm going to have questions.

The last time I tried to work with a frame that had a terminated string (the description on the comment/lyrics/user-text frames) you had to clean up after me. So I'm a little leery messing with these, and with how to implement the "even-odd" aspect of the `.version4` spec.

SDGGiesbrecht commented:
On the parsing end, just use `extractPrefixAsStringUntilNullTermination` repeatedly until the data is empty. On the writing end, just add all the strings one after another (although I cannot tell from the specification whether the very last string should be null‐terminated or not).

NCrusher74 commented:
Okay. Do you think that will apply to `.version4` as well?

NCrusher74 committed:
added credits frame type

SDGGiesbrecht commented:
Yes. It sounds the same. Maybe like version 4 has many fields like that, so the details were centralized elsewhere in the specification?

NCrusher74 commented:
Just what's described under the "MusicianCredits" frame, which I included there as well. They're the only two frames like that in `.version4` as far as I know.

NCrusher74 committed:
Added InvolvedPeopleList and MusicianCredits Frame Creators

NCrusher74 committed:
added CreditsList type frame creator, with both strings terminated.

NCrusher74 committed:
added frame creators to frame creator chain

NCrusher74 committed:
Okay clearly I'm not doing the frame creator encoding/termination correctly.

NCrusher74 commented on Source/Creation/Frame/Content/ID3FrameCreatorsChainFactory.swift:
```diff
@@ -51,8 +51,15 @@ class ID3FrameCreatorsChainFactory {
+        let frameFromCreditsListISO88591ContentCreator = ID3CreditsListFrameCreator(
+            frameContentSizeCalculator: frameContentSizeCalculator,
+            frameFlagsCreator: frameFlagsCreator,
+            stringToBytesAdapter: ID3ISO88591StringToByteAdapter(
```
Am I using the correct StringToBytes adapter here? Because the test for this frame creator is coming out strangely. (see below.)

SDGGiesbrecht replied:
Well, `ISO88591` matches between the names `frameFromCreditsListISO88591ContentCreator` and `ID3ISO88591StringToByteAdapter`. So if something is wrong, the problem is elsewhere.

NCrusher74 commented on Tests/Creation/Frame/Content/String/ID3CreditsListFrameCreatorTest.swift:
```diff
@@ -0,0 +1,58 @@
+    }
+}
+
+/* XCTAssertEqual failed: ("[
```
This test is coming out all sorts of wonky. It seems to be expecting a null byte between each character?

SDGGiesbrecht replied:
You’ve hooked it to a `ID3UTF16StringToByteAdapter`. UTF‐16 uses two bytes (16 bits) for each character. In the ASCII range, the first half is a zero byte and the second half is the familiar byte.

NCrusher74 committed:
updated README with new frames and proper credit

NCrusher74 committed:
alphabetize ALL THE THINGS

NCrusher74 commented:
I’m on my phone at the moment I copied the test from the user defined text
test, I believe. I noticed it used that utf16 string adapter but iirc it
uses ISO88591 here. I had noted it but since it was working for the User
Text frame creator test I went with it. Why does it work for user text but
not for my test? (Not questioning that I did it wrong, just trying to
understand why.)

On Thu, Apr 2, 2020 at 1:00 PM Jeremy David Giesbrecht <
notifications@github.com> wrote:

> *@SDGGiesbrecht* commented on this pull request.
> ------------------------------
>
> In Source/Creation/Frame/Content/ID3FrameCreatorsChainFactory.swift
> <https://github.com/NCrusher74/ID3TagEditor/pull/5#discussion_r402576597>:
>
> > @@ -51,8 +51,15 @@ class ID3FrameCreatorsChainFactory {
>                  paddingAdder: paddingAdder,
>                  frameConfiguration: frameConfiguration)
>          )
> -
> -// MARK: Specific Frame Creators (Text)
> +        let frameFromCreditsListISO88591ContentCreator = ID3CreditsListFrameCreator(
> +            frameContentSizeCalculator: frameContentSizeCalculator,
> +            frameFlagsCreator: frameFlagsCreator,
> +            stringToBytesAdapter: ID3ISO88591StringToByteAdapter(
>
> Well, ISO88591 matches between the names
> frameFromCreditsListISO88591ContentCreator and
> ID3ISO88591StringToByteAdapter. So if something is wrong, the problem is
> elsewhere.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/NCrusher74/ID3TagEditor/pull/5#discussion_r402576597>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANRMHJAHQWLORQIN2QMN3F3RKTVFVANCNFSM4LZUYRSQ>
> .
>

SDGGiesbrecht commented:
You would have to point me at whatever place you mean. “test” in Latin‐1 is `0x74, 0x65, 0x73 0x74`, but in UTF‐16 it’s `0x00, 0x74, 0x00, 0x65, 0x00, 0x73, 0x00, 0x74` (because it’s really `0x0074, 0x0065, 0x0073, 0x0074`).

SDGGiesbrecht commented:
You would have to point me at whatever place you mean. “test” in Latin‐1 is `0x74, 0x65, 0x73 0x74`, but in UTF‐16 it’s `0x00, 0x74, 0x00, 0x65, 0x00, 0x73, 0x00, 0x74` (because it’s really `0x0074, 0x0065, 0x0073, 0x0074`).

NCrusher74 commented:
Sorry I’m still on my phone here. Maybe it wasn’t a user text yes but a
comment types test. Basically it was a test in the same folder as this test
and the StringURL frame creator test that I made a few days ago, but it was
for the frame creator you worked on.

On Thu, Apr 2, 2020 at 2:02 PM Jeremy David Giesbrecht <
notifications@github.com> wrote:

> You would have to point me at whatever place you mean. “test” in Latin‐1
> is 0x74, 0x65, 0x73 0x74, but in UTF‐16 it’s 0x00, 0x74, 0x00, 0x65,
> 0x00, 0x73, 0x00, 0x74 (because it’s really 0x0074, 0x0065, 0x0073, 0x0074
> ).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/NCrusher74/ID3TagEditor/pull/5#issuecomment-608082840>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANRMHJC24TIRFOUAHLMH5OTRKT4ONANCNFSM4LZUYRSQ>
> .
>

NCrusher74 commented:
Okay, sorry, long day at work, I'm finally home. I was referring to `ID3CommentTypesFrameCreatorTest` as the test that I modeled this test after. It also uses `ID3UTF16StringToByteAdapter` but in `ID3FrameCreatorsChainFactory`, there's this:

```
        let commentFrameCreator = ID3CommentFrameCreator(
            frameCreator: frameFromMultiStringISO88591ContentCreator,
            id3FrameConfiguration: frameConfiguration
        )
```

```
        let frameFromMultiStringISO88591ContentCreator = ID3CommentTypesFrameCreator(
            frameContentSizeCalculator: frameContentSizeCalculator,
            frameFlagsCreator: frameFlagsCreator,
            stringToBytesAdapter: ID3ISO88591StringToByteAdapter(
                paddingAdder: paddingAdder,
                frameConfiguration: frameConfiguration)
        )
```

Unless...is that an artifact from when I screwed up the CommentTypes frame that didn't get deleted when you fixed the rest?

NCrusher74 committed:
Added Content Parsing Operation. Frame creator test still not passing.

NCrusher74 commented on Tests/Creation/Frame/Content/String/ID3CreditsListFrameCreatorTest.swift:
```diff
@@ -17,7 +17,7 @@ class ID3CreditsListFrameCreatorTest: XCTestCase {
             ),
             frameFlagsCreator: ID3FrameFlagsCreator(),
-            stringToBytesAdapter: ID3UTF16StringToByteAdapter(
+            stringToBytesAdapter: ID3ISO88591StringToByteAdapter(
```
I changed the StringToByteAdapter, and now the failure message is:

```
XCTAssertEqual failed: ("[34, 0, 0, 0, 25, 0, 0, 0, 109, 117, 115, 105, 99, 105, 97, 110, 0, 0, 77, 117, 115, 105, 99, 105, 97, 110, 32, 78, 97, 109, 101, 0]") 
is not equal to ("[34, 0, 0, 0, 52, 0, 0, 1, 109, 117, 115, 105, 99, 105, 97, 110, 0, 77, 117, 115, 105, 99, 105, 97, 110, 32, 78, 97, 109, 101, 0]")
```

So it's expecting another null byte between the strings?

Which is fine, I can rewrite the test to include that, but I'd like to understand why so I'm not just rigging the test to pass without making sure I'm actually doing it right.

SDGGiesbrecht replied:
I don’t think there should be two 0 bytes unless the encoding is 16‐bit. (In 8‐bit, null is `0x00`; in 16‐bit, it is `0x0000`.) Without looking at the code, there are (at least) two possibilities:

1. The string encoding method is terminating the string with null automatically, and then you terminating it with another null manually.
2. The string encoding method has its wires crossed and is encoding the string in 8‐bit, but the termination null in 16‐bit.

NCrusher74 replied:
For the most part, I'm imitating what you did with the UserDefinedTextInformation frame, or the CommentTypes frame since you used that type for the UserText frame in some places.

Where would I look to see if this is happening?

This is the CreditsListFrameCreator:

```swift
class ID3CreditsListFrameCreator: CreditsListFrameCreator {
    
    private let frameContentSizeCalculator: FrameContentSizeCalculator
    private let frameFlagsCreator: FrameFlagsCreator
    private let stringToBytesAdapter: StringToBytesAdapter
    
    init(frameContentSizeCalculator: FrameContentSizeCalculator,
         frameFlagsCreator: FrameFlagsCreator,
         stringToBytesAdapter: StringToBytesAdapter) {
        self.frameContentSizeCalculator = frameContentSizeCalculator
        self.frameFlagsCreator = frameFlagsCreator
        self.stringToBytesAdapter = stringToBytesAdapter
    }
    
    func createFrame(
        frameIdentifier: [UInt8],
        version: ID3Version,
        role: String,
        person: String
    ) -> [UInt8] {
        let frameContents = [UInt8](
            [
                stringToBytesAdapter.adapt(
                    string: role,
                    for: version,
                    includingEncoding: true,
                    includingTermination: true
                ),
                stringToBytesAdapter.adapt(
                    string: person,
                    for: version,
                    includingEncoding: true,
                    includingTermination: true
                )
                ].joined()
        )
        
        return [UInt8](
            [
                frameIdentifier,
                frameContentSizeCalculator.calculateSizeOf(
                    content: frameContents, version: version),
                frameFlagsCreator.createFor(version: version),
                frameContents
                ].joined()
        )
    }
}
```

So it does include the termination on both strings, which is from your (changes to?) the StringToBytesAdapter. Since this is only testing the FrameCreator, I don't think anything I included in the Parsing (which I'm a lot iffier on) could be affecting it, could it?

Everything I can see for the encoding is ISO8859-1. It's using `UInt32ToByteArrayAdapterUsingUnsafePointer` like the test I am emulating. Could that be an issue?

SDGGiesbrecht replied:
Actually the first of the pair of nulls is the termination of the preceding string, and the second (wrongly) is the encoding of the following string (Latin 1). Only the very first string in the list should include the encoding specifier. The rest should use `includingEncoding: false`.

NCrusher74 replied:
Ah ok. Thank you!

NCrusher74 replied:
Okay, this next part is the tricky bit. Per the spec, I've got the `role` and `person` strings in place, each terminated.

Now I need to be able to add more `role`:`person` pairs *on the same frame*, because the spec (at least for `.version3`, says there should only be one IPLS frame per tag.

So, on the API end, ideally I think it should be just like adding a new frame, but behind the scenes, what is actually happening is that the strings are being appended to the existing frame, such that the frame reads `role1`,`person1`,`role2`,`person2`,`role3`,`person3, person4, person5` (since the `person` string should be able to be a comma delimited string with multiple people fulfilling the same role, per the `.version4` spec.)

And on the parsing end, it should basically come back as though there were multiple `IPLS/TIPL/TMCL` frames rather than just one, with each `role`:`person` pair being returned separately.

So this is different from, for example, the `TXXX` frame where there is more than one frame allowed per tag as long as each one has a different description (which is still an issue I need to tackle in `ID3TagEditor`.)

I suspect in the end the code isn't going to be that complicated, but I have been looking at it since last night scratching my head because I can't figure out where this functionality should go? I'm just drawing a blank on how to get started with it.

NCrusher74 replied:
I'd ask if maybe the `role:person` pair should be handled as a dictionary entry instead of separate strings, but that is sort of made impossible by the fact that there may be multiple people fulfilling the same role.

SDGGiesbrecht replied:
Start by making the API that just does exactly like the specification says and make sure it works. (Then you can add your own convenience methods on top of that.)

That means, for starters, instead of these parameters...

```swift
role: String,
person: String
```

...you need this...

```swift
entries: [(role: String, person: String)]
```

NCrusher74 replied:
Thank you! Will do!

NCrusher74 committed:
Added frame creator tests for specific tags. Stuck until I can figure out how to append "role:person" pairs to existing frame as Version3 spec says there should only be one IPLS frame per tag.

So there should be one frame, that returns multiple pairs (if used) which should be parsed as though each pair were on a separate frame.

NCrusher74 committed:
Frame creation tests passing.

NCrusher74 committed:
Init not working properly yet. Will work on it more after work.

NCrusher74 committed:
getting closer. Now I really need to go to work.

NCrusher74 committed:
I cannot figure out how to write the credits list frame creator

NCrusher74 commented on Source/Creation/Frame/Content/String/ID3CreditsListFrameCreator.swift:
```diff
@@ -29,28 +29,30 @@ class ID3CreditsListFrameCreator: CreditsListFrameCreator {
-                        includingTermination: true
-                    )
-                    )]
+                stringToBytesAdapter.encoding(for: version),
```
I've been trying to figure this out for hours, but I can't work out how to write this function:

```swift
    func createFrame(
        frameIdentifier: [UInt8],
        version: ID3Version,
        entries: [(role: String, person: String)]
    ) -> [UInt8] {
        let frameContents = [UInt8](
            [
                stringToBytesAdapter.encoding(for: version),
                entries: [
                stringToBytesAdapter.adapt(
                string: role,
                for: version,
                includingEncoding: true,
                includingTermination: true
                ),
                stringToBytesAdapter.adapt(
                string: person,
                for: version,
                includingEncoding: false,
                includingTermination: true
                )
                ]
                ].joined()
        )
        
        return [UInt8](
            [
                frameIdentifier,
                frameContentSizeCalculator.calculateSizeOf(
                    content: frameContents,
                    version: version),
                frameFlagsCreator.createFor(version: version),
                frameContents
                ].joined()
        )
    }
```

SDGGiesbrecht commented:
You’ll need a `for` loop. Something like this:

```swift
var frameContents: [UInt8] = []
for entry in entries {
  // Add the bytes for the entry’s role and person into “frameContents”
}

// The last part is the same:
return [UInt8](
  [
    frameIdentifier,
    frameContentSizeCalculator.calculateSizeOf(
      content: frameContents,
      version: version),
    frameFlagsCreator.createFor(version: version),
    frameContents
  ].joined()
)
```

NCrusher74 commented:
Okay., thank you.

Do I need the `stringToBytesAdapter.encoding(for: version)` still?

SDGGiesbrecht commented:
The _first_ one needs a string encoding flag. You can either track which one is first, or you can encode the string encoding flag separately before starting the loop, and then encode all entries without it.

NCrusher74 commented:
I think I've almost got it except for one small, silly thing:

```swift
    func createFrame(
        frameIdentifier: [UInt8],
        version: ID3Version,
        entries: [(role: String, person: String)]
    ) -> [UInt8] {
        var frameContents:[UInt8] = []
        frameContents.append(contentsOf: stringToBytesAdapter.encoding(for: version))

        for entry in entries {
            var rolePersonPair: [UInt8] = []
            let adaptedRole = stringToBytesAdapter.adapt(
                string: role,
                for: version,
                includingEncoding: false,
                includingTermination: true
            )
            let adaptedPerson = stringToBytesAdapter.adapt(
                string: person,
                for: version,
                includingEncoding: false,
                includingTermination: true
            )
            rolePersonPair.append(contentsOf: adaptedRole)
            rolePersonPair.append(contentsOf: adaptedPerson)
            frameContents.append(contentsOf: rolePersonPair)
        }
        
        return [UInt8](
            [
                frameIdentifier,
                frameContentSizeCalculator.calculateSizeOf(
                    content: frameContents,
                    version: version),
                frameFlagsCreator.createFor(version: version),
                frameContents
                ].joined()
        )
    }
```

In the 
```swift
            let adaptedRole = stringToBytesAdapter.adapt(
                string: role,
                for: version,
                includingEncoding: false,
                includingTermination: true
            )
```
portions, it's not recognizing the identifiers `role` and `person`. Which...shouldn't it be getting those from the parameter `entries: [(role: String, person: String)]`?

I mean, I could add empty string variables for `role` and `person` but they're already declared in the parameter?

NCrusher74 commented:
I suspect the above is because I'm not declaring the parameters correctly in the initializer?

```swift
public class ID3FrameCreditsList: ID3Frame, Equatable, CustomDebugStringConvertible {
    /// An array of the role:person tuples
    public var entries: [(role: String, person: String)]
//    /// The role the involve person performs
//    public var role: String
//    /// The involved person or performer
//    public var person: String
    /// CreditsList description, useful for debug.
    public var debugDescription: String {
        return "entries"
    }
    
    /**
     Init an ID3 involved person and musician credits frame.
     - parameter entries: the array of `role`:`person` pairs
     - parameter role: the role of the involved person.
     - parameter person: the name (or comma-delimited names) of the person fulfilling a given role.
     */
    public init(entries: [(role: String, person: String)]) {
        self.entries = entries
//        self.role = role
//        self.person = person
    }
    
    /**
     Compare two roles.
     
     - parameter lhs: left side of compare operation.
     - parameter rhs: right side of compare operation.
     
     - returns: true if the role values are the same, else false.
     */
    public static func ==(lhs: ID3FrameCreditsList, rhs: ID3FrameCreditsList) -> Bool {
        return lhs.role == rhs.role
    }
}
```

I'm struggling to figure out the proper way to declare parameters that are nested like this.

SDGGiesbrecht commented:
> it's not recognizing the identifiers `role` and `person`.

`entry.role` and `entry.person`. Each `entry` is a [named tuple](https://docs.swift.org/swift-book/LanguageGuide/TheBasics.html#ID329).

NCrusher74 commented:
... I could have sworn I tried that at one point.

Maybe I did but something else was wrong and I changed it thinking it would fix the other problem. Thank you.

NCrusher74 commented:
Okay, I think the only other part I need to get right to make the frame creation side build is this:

```swift
    /**
     Compare two roles.
     
     - parameter lhs: left side of compare operation.
     - parameter rhs: right side of compare operation.
     
     - returns: true if the role values are the same, else false.
     */
    public static func ==(lhs: ID3FrameCreditsList, rhs: ID3FrameCreditsList) -> Bool {
        return lhs.role == rhs.role
    }
```

I'm not even sure if this is the proper way to go about checking for this, since it's intended to prevent two frames of the same type, so maybe I should just go with `entries`? (except that (role: String, person: String) isn't equatable?)

But somewhere along the way I need to put in filtering so that, if there are identical `role`s, they get turned into a single `role` with a comma-delimited string for `person`

SDGGiesbrecht commented:
That is part of [the `Equatable` protocol](https://developer.apple.com/documentation/swift/equatable). Unless upstream is misusing it, two frames should be considered equal only if they actually are completely equal. While equality excludes irrelevant things like cache state, all properties inherent to the meaning of the thing should be take into account. I believe in this case that means all the roles, all the people, and their order are important. So simply using `==` to compare the two arrays ought to do it succinctly.

NCrusher74 commented:
I think perhaps I've been misusing it, then, because for something like the comment frame type, `description` and `language` are taken into account, but not `content`?

For this particular frame, perhaps I'm writing the initializer wrong.

```swift
    public init(entries: [(role: String, person: String)]) {
        self.entries = entries
    }
```

Because:

```swift
    public static func == (lhs: ID3FrameCreditsList, rhs: ID3FrameCreditsList) -> Bool {
        return lhs.entries == rhs.entries
    }
```
gets me:

"Type '(role: String, person: String)' cannot conform to 'Equatable'; only struct/enum/class types can conform to protocols"

But any attempt to write it differently (for instance, declaring `role` and `person` as variables and adding them into the initializer) causes other errors.

SDGGiesbrecht commented:
Yes. Sorry. I forgot that tuples aren’t actually `Equatable`, even though they have the `==` method for it, hence the `Array: Equatable where Element: Equatable` does not include them.

Does this work?

```swift
lhs.entries.elementsEqual(rhs.entries, by: { $0 == $1 })
```

NCrusher74 commented:
Yes, that worked, thank you!

The frame creators seem to be working and the tests passing. Now time for the parsing!

NCrusher74 committed:
Frame creator tests passing.

NCrusher74 committed:
Can't figure out how to write acceptance tests

NCrusher74 commented:
Okay, sort of related to the same issue as last night. I can't figure out how, formatting this frame, the way it is, isolating the strings to check their accuracy should work.

This is how it's done for `UserDefinedTextInformation`:

```swift
        XCTAssertEqual((id3Tag?.frames[.UserDefinedUrl] as? ID3FrameUserDefinedText)?.description, "UserDefinedUrl")
        XCTAssertEqual((id3Tag?.frames[.UserDefinedUrl] as? ID3FrameUserDefinedText)?.content, "http://UserDefinedUrl.com")
```

But this gets me the same error about not being about to use `equatable`

```swift
        XCTAssertEqual((id3Tag?.frames[.InvolvedPeople] as? ID3FrameCreditsList)?.entries, [("Producer", "Producer Name"), ("Director", "Director Name")])
```

I've been trying to adapt the code used for the duplicate frame checking that you recommended above, but I'm not getting it right.

Also, I'm not sure `lhs.entries.elementsEqual(rhs.entries, by: { $0 == $1 })` is working for its intended purpose? Or maybe it is and I just found a way to screw up anyway?

It builds, but I discovered this morning that in the `FrameCreatorChainFactory` I'd accidentally been writing two of the `InvolvedPeople` frames instead of one `InvolvedPeople` frame and one `MusicianCredits` frame due to a cut and paste error.

Instead of 
``` swift       
let musicianCreditsListFrameCreator = ID3MusicianCreditsListFrameCreator(
frameCreator: frameFromCreditsListISO88591ContentCreator,
id3FrameConfiguration: frameConfiguration
```

I had:

```swift
let musicianCreditsListFrameCreator = ID3InvolvedPeopleListFrameCreator(
frameCreator: frameFromCreditsListISO88591ContentCreator,
id3FrameConfiguration: frameConfiguration
```

And so what I was getting was a file with duplicate `InvolvedPeople` frames.

Isn't that something that should be caught at some point, if `ID3TagEditor` is going to be working to spec?

NCrusher74 commented:
I wonder if, instead of writing by [Frame Name: Frame Type] what we need to do is create a way of checking if the frames are allowed by the spec to be duplicated or if they have to be unique, and write by that?

SDGGiesbrecht commented:
You’ll need a `FrameKey` type that is a frame name _and_ whatever other information it needs to uniquely identify its “slot” in the tag (such as the language of a comment frame). Then `FrameKey` would be used as the dictionary key instead.

NCrusher74 commented:
Yeah, that's what I was thinking. I do still need to write upstream about it and make sure that's okay with him.

NCrusher74 committed:
just need to get the acceptance test working

NCrusher74 commented:
Okay, I think I worked out how to write the Acceptance reader test to get around the equatable issue, but I ended up with a fatalError in FrameParser.swift when running it because I hadn't added them to the `threeByteString`/`fourByteString` enums.

Which is easy enough to fix for the `threeByteString`, but the `fourByteString` changes between `.version3` and `.version4`. `.version3` is "IPLS" and `.version4` is "TIPL".

I'm trying to add them , but I think the fact that the version enum is `UInt8` is throwing me off. I've tried just about every variation of this that I can think of:

```swift
    internal var fourByteString: String? {
        switch self {
            case .Comment:
                return "COMM"
            case .InvolvedPeople:
                let version3 = ID3Version(rawValue: 3)
                let version4 = ID3Version(rawValue: 4)
                if ID3Version == .version3 {
                    return "IPLS"
                } else if ID3Version == .version4 {
                    return "TIPL"
            }
            case .UnsyncedLyrics:
                return "USLT"
            case .UserDefinedTextInformation:
                return "TXXX"
            case .UserDefinedUrl:
                return "WXXX"
            default:
                return nil
        }
    }
    ```

but I'm not finding the correct way to write it.

SDGGiesbrecht commented:
There is no version available to that property yet. You’ll have to add a parameter. If you do that, you may as well merge the three and four byte properties:

```swift
func identifier(version: ID3Version) -> String? {
  switch self {
  case .Comment:
    switch version {
      case .version2:
        return "COM"
      case .version3, version4:
        return "COMM"
    }
  case .InvolvedPeople:
  // ...
  }
}
```

NCrusher74 commented:
Okay will do!

NCrusher74 commented:
But before I do, just so I understand how your changes work versus the original functionality of `ID3TagEditor`, what is the difference in purpose between your three/four byte string property, and the way the id3Identifiers are listed in [ID3FrameConfiguration.swift](https://github.com/NCrusher74/ID3TagEditor/blob/master/Source/Frame/ID3FrameConfiguration.swift)?

SDGGiesbrecht commented:
They serve exactly the same purpose.

I hadn’t found the `ID3FrameConfiguration.swift` yet when I first wrote those properties. (Methods and properties of the type you are dealing with (`FrameName`) show up in code completion; unrelated types like `ID3FrameConfiguration` do not.) By the time I had stumbled across it, I didn’t care to convert what I’d written back over to their way because (in addition to being isolated and hard to find):

1. A property with a `switch` statement is better than a static dictionary. If they were all in the switch statement instead of the dictionary (and without a `default` case), the compiler would tell you if one were missing. The dictionary won’t fail until runtime, and with those `!` marks it would just crash helplessly.
2. Re‐initializing an entire configuration class instance with all those dictionaries every time you want one little piece of information is a waste of memory and CPU time. (And avoiding that by passing around the same instance to and from every method is an unnecessary headache when the information itself is fixed and unalterable.)
3. Everything about the `enum` and its methods is static. The compiler can optimize them down to almost nothing. But a `class` with dictionary properties is a big opaque thing that the compiler cannot prove won’t be altered behind its back at runtime, so there is little optimization it can do.

But upstream seems very opinionated about how the code is arranged, so it was one of those things I figured they might ask you to reorganize their way before merging.

NCrusher74 commented:
Okay, once I merged the three/four byte properties into an "identifier" method, I edited `extractHeader` in `FrameParser` as follows:

```swift
    internal func extractHeader(from frameData: inout Data.SubSequence, version: ID3Version) {
        let name = frameData.extractFirst(version.identifierSize)
        assert(
            String(ascii: name) == frameName.identifier,
            "Mismatched frame name: \(String(ascii: name)) ≠ \(frameName.identifier!)"
        )
        _ = frameData.extractFirst(version.sizeSize)
        _ = frameData.extractFirst(version.flagSize)
    }
```

which gets me an "Type '(ID3Version) -> String?' cannot conform to 'StringProtocol'; only struct/enum/class types can conform to protocols" error. 

And as I was writing this, your last response came in, and it begs the question...At what point do our changes become so divergent from what upstream intended that it's not worth his or our while to do a pull request and if so, what is the accepted protocol in that situation? Especially if he isn't really supporting it himself anymore?

(This brought to you by the fact that I keep looking at how each and every frame has its own frame creator type, and how much larger this package is than, say, `TagLibKit` and scratching my head, thinking SURELY there's a more streamlined way of handling things.)

SDGGiesbrecht commented:
> "Type '(ID3Version) -> String?' cannot conform to 'StringProtocol'

`frameName.identifier` needs to be `frameName.identifier(version: version)`.

> At what point do our changes become so divergent from what upstream intended that it's not worth his or our while to do a pull request and if so, what is the accepted protocol in that situation? Especially if he isn't really supporting it himself anymore?

🤷‍♂ If you would rather query those dictionaries, that’s fine. You could also just remove the `assert`. It was helpful during debugging, but it gets stripped out when compiling for release. See [its documentation](https://developer.apple.com/documentation/swift/1541112-assert).

NCrusher74 commented:
Actually, I was thinking more along the lines of doing away with those dictionaries and doing it all your way. But I'm not sure that's...polite? Acceptable? idk.

NCrusher74 commented:
Okay, this is how I worked around the issue with `equatable` and tuples, but the `XCTAssertTrue` test is failing, so I'm stumped.

```swift
        let involvedPeopleArray = [("Producer", "Producer Name"), ("Director", "Director Name")]
        let involvedPeopleFrame = (id3Tag?.frames[.InvolvedPeople] as? ID3FrameCreditsList)?.entries ?? []
        XCTAssertTrue(involvedPeopleArray.elementsEqual(involvedPeopleFrame, by: { $0 == $1 }))
```

SDGGiesbrecht commented:
Break this line down and see where it diverges from your expectations:
```swift
let involvedPeopleFrame = (id3Tag?.frames[.InvolvedPeople] as? ID3FrameCreditsList)?.entries ?? []
```

NCrusher74 commented:
Hrm.

```swift
        print(id3Tag?.frames[.InvolvedPeople])
        print(id3Tag?.frames[.InvolvedPeople] as? ID3FrameCreditsList)
        print((id3Tag?.frames[.InvolvedPeople] as? ID3FrameCreditsList)?.entries)
```
gets me:
```
Optional([(role: "Producer", person: "Producer Name")])
Optional([(role: "Producer", person: "Producer Name")])
Optional([(role: "Producer", person: "Producer Name")])
```

Which has to be a problem with the parsing, because that frame is written properly. This is what I see when I look at it in Yate:

<img width="422" alt="Screen Shot 2020-04-05 at 3 54 20 PM" src="https://user-images.githubusercontent.com/56804260/78512023-0a2f8280-7756-11ea-9ebd-4c81768b5d03.png">

NCrusher74 commented:
This is my `ID3FrameCreditsListParsingOperation`:

```swift
struct ID3CreditsListFrameContentParsingOperation: FrameContentParsingOperation, FrameParser {
    
    let frameName: FrameName
    
    func parse(frame: Data, version: ID3Version, completed: (FrameName, ID3Frame) -> ()) {
        var parsing = frame[...]
        extractHeader(from: &parsing, version: version)
        
        let encoding = extractEncoding(from: &parsing)
        
        let parsed = extractCreditStrings(from: &parsing, encoding: encoding)
        let constructed = ID3FrameCreditsList(
            entries: [(role: parsed.role ?? "",
                       person: parsed.person ?? "")]
        )
        completed(frameName, constructed)
    }
}
```

So I suppose what I need is to work in `for entry in entries` somehow, but I'm not sure exactly how to do that because it's using `(entries: [role:person:])` as the parameter for `ID3FrameCreditsList.` So that is what I need to work out.

SDGGiesbrecht commented:
`extractCreditStrings(from:encoding:)` ought to return an array, no?

Inside that function, you would do something like this:
```
var strings: [String] = []

while !parsing.isEmpty,
  let next = parsing.extractPrefixAsStringUntilNullTermination(encoding) {
  strings.append(next)
}
```

And then you would pair the strings in the array using any of a dozen strategies. [Here is just one.](https://stackoverflow.com/questions/34550630/iterate-over-collection-two-at-a-time-in-swift)

NCrusher74 commented:
That's what I'm trying to do now, though of course what you're recommending is much cleaner than the messy ways I have been trying. Thank you!

NCrusher74 committed:
almost there.

NCrusher74 commented:
I went with the Array.pairs() extension and added it to the convenience extensions.

I've got to run out the door to work now, but I think I'm almost there. I just need to figure out how to reconcile the fact that sometimes I've got functions expecting [(role: String, person: String)] as a parameter or return, and other times I've got [(String, String)] and sometimes a function that is expecting the former is receiving the latter. I'll work on that when I get home tonight.

NCrusher74 committed:
parser working. Tests passing.

NCrusher74 commented:
Considering other apps are reading these frames perfectly the way they are, I think I'm going to scrap any ideas I had about having it return the frames in a different way and leave them as is.

Which means...it's time to merge this branch and move on to the final phase: The CTOC and CHAP frames. *eep*
